### DISCUSSION TITLE
Sending key presses to a specific open Window with xdotool

### DISCUSSION THREAD

**Discussion by Danie10** [streamdeck-linux-gui/streamdeck-linux-gui#56] [Show and tell]
Seeing the recent update to include the use of num pad keys I thought I'd show how I was doing that, and also ensuring the key presses go to my OBS window, without losing focus of what is currently active (global Press Keys would achieve the same keeping active focus). This works with X11 and I don't think with Wayland.

I use these shortcuts in OBS to zoom in on areas of the screen whilst I'm working on my other monitor. Sometimes that includes typing in a terminal screen, so I don't want any global key presses to show anything in that terminal (hence directing the key presses to the OBS window. I have the button images set to show arrows pointing to the various directions from the centre button.

The text below is pasted into the **command field** for the relevant button, and requires that xdotool is installed (it simulates key presses and mouse clicks, and can move and resize windows). The 'search --name' parameter searches for the first window that is open, and that starts with the name OBS x.x.x and where the x's represent any numeric version numbers (this is how OBS windows are named eg. OBS 3.4.1). The key parameter is usually the key abbreviations like ctrl+v but can also include keysym hex codes for more difficult keys such as the Alt Left key being 0xffe9 (or key Alt_L) - see more at https://gitlab.com/cunidev/gestures/-/wikis/xdotool-list-of-key-codes.

`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Home` for HOME or numeric 7 key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Up` for up arrow or numeric 8 key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Prior` for PgUp or 9 numeric key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Left` for left arrow or 4 numeric key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Begin` for middle or numeric 5 key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Right` for right arrow or 6 numeric key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_End` for End or numeric 1 key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Down` for down arrow or 2 numeric key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Next` for PgDn or 3 numeric key
`xdotool search --name '^OBS [0-9]+\.[0-9]+\.[0-9]' key ctrl+KP_Multiply` for * key

You can also add delays between key presses, chain them together, include scripts, and lots more. So xdotool could execute say ctrl+f then pause for 1 second and then press ctrl+shift+t for you.

Another trick is to use xdotool to say refresh your browser's tab number 3 (pin them to keep it consistent), where searching for the window name of your browser will give it focus for the keypress, and ctrl+3 will focus on tab 3 for further key actions (ctrl+L focusses on the URL address bar).

To find open window names run `wmctrl -l | grep -o "$HOSTNAME.*" | sed "s/$HOSTNAME //g"`

You can use the `xev` tool and capture the key you are looking for - just execute xev in the terminal. In the output, look for the **keysym hex value**, for example `(keysym 0x74, t)`

**Comment by oparada1988** [comment]
oh cool! Thanks for the guide and explination. Will definetly be setting this up now!

**Reply by Danie10** [reply]
Thanks, yes, I realised there is tremendous power in xdotool for some things we may want to do already. Others can add additional related tips in the discussion below, too.
